discussion/rust-for-embedded-systems.tex

```diff
@@ -0,0 +1,90 @@
+
+In \autoref{sec:intro:assignment} we identified six language challenges that must be considered when using a language in an embedded system.
+Here, we revisit each challenge and discuss how they were solved for the {\rg} platform.
+
+\begin{description}
+\item [LC1 {\lci}] \hfill \\
+  {\rust} exposes two \concept{intrinsics} for handling volatile read and write.
+  These makes the code more verbose compared to the mechanism in {\C}, where a variable is marked with the \code{volatile} keyword.
+  In {\rust}, the intrinsic functions must be used each time a variable is read or written.
+  This, however, gives the programmer more fine-grained control as a variable can be used both as volatile and non-volatile.
+
+\item [LC2 {\lcii}] \hfill \\
+  Interrupt handlers must use the {\C} \gls{abi}, because the interrupts are dispatched from the {\C} runtime.
+  This makes the code more verbose, but it is easily achievable in {\rust}.
+  As discussed in \autoref{sec:avoiding_mutable_aliases_to_hardware}, a downside to the interrupted programming model found in embedded programming for {\rust} is the reliance on global mutable state.
+  In these circumstances, the compiler is limited in its ability to statically verify safety because accesses to these variables must be contained inside {\unsafe} blocks.
+
+\item [LC3 {\lciii}] \hfill \\
+  In this thesis, we have only considered hardware devices as \glspl{mmio}.
+  As {\rust} supports raw pointers and allows the programmer to access arbitrary memory addresses and cast these as structs, the handling of hardware registers are equally practical in {\rust} as in {\C}.
+  Given a more mature embedded platform in {\rust}, we foresee that even more of the compile-time ownership analysis provided by the {\rust} compiler can be used to ensure safe interactions with hardware registers.
+
+\item [LC4 {\lciv}] \hfill \\
+  {\rust} (and {\C}) have, unlike many other programming languages, no life before the {\main} function.
+  In this statement lies the fact that in the global scope, one can only initialize objects which have constant initializers.
+  Therefore, all the initializers only contain constant data and they can be handled by the startup mechanisms described in \autoref{sec:back:startup}.
+  This implies that the static object construction problem is non-existent in {\rust} programs.
+
+\item [LC5 {\lcv}] \hfill \\
+  Dynamic allocation in {\rust} is implemented in the \lib{alloc} library.
+  We were easily able to include this functionality in \gls{rel}.
+  In \autoref{sec:res:heap}, we showed that the heap allocation in {\rust} and {\C} are identical.
+  This is due to the allocation algorithm in {\rust} being directly dependent on the \lib{newlib} \func{malloc} implementation, and therefore the memory fragmentation is equal to that of the existing {\C} platform.
+  Note that this does not hold true for the regular {\rust} platform because it uses a different allocation algorithm.
```

Comment:
`Note that this does not strictly hold true for all {\rust} platforms because the can use different allocation algorithms.`